### PR TITLE
Feature/exampleapp

### PR DESCRIPTION
**Pull request by ingkor**
Reworked to controllers. 

There is some additional work needed: 
I started a domainevent implementation, but have not tested it or completed it.
Some guidance or insights is welcome.

ToDo:
-Finish domainevent implementation
-Check naming conventions and openapi conventions (endpoint name for instance)
-Add Timer on calls to see performance of specification pattern + add to response object
-Look into compiled queries of EF core and compare to specification performance
-Front-end implementation? Swagger is already implemented for the backend devs.

### CODE REVIEW

**Comment by ingkor** [comment]
@vittorelli @Niels1986  @Clivar v1 of exampleapp is done.
Please check locally if it works.

Added in this PR:
`Vivarni.Example.API`

- OpenAPI implemented with 2 controllers. For `GuestMessage` and `GuestmessageCounter` entities.
- Domainevent adds a GMC when a GM gets added.
- ApiModels implemented using Records
- SQLLite db in API

`Vivarni.Example.Domain
`
- 2 entities (GM, GMC) 
- `DomainEvent` and `DomainEventHandler` on GM 
- Specification for GM on date (not used right now)

`Vivarni.Example.Infrastructure`

- `ApplicationDbContext`
- Migrations for the SQLLite db


On to do for v2 (and later versions):

- Front end communicating with api, mobile or web (with oauth in swagger and FE?)
- Rework domainevents to create a meaningful domainevent visually represented in frontend
- Implement specification & remaining methods in the genericrepository
- Restructure project => could use a guideline where to put which parts of the implementation: e.g. domainevents, domaineventhandlers
- Extend readme with a bigger introduction and guideline. Links to existing repos using Vivarni?

**Comment by Clivar** [comment]
I modified it since I expect domainevents to be handled in the application layer in general.
But I see now why you wrote it this way. A clever way to provide DI for methods on domain entity objects.
I am not against it. I believe DomainEvents are maybe not supposed to be exclusively handled in the application layer.

@vittorelli What do you think?

**Comment by ingkor** [comment]
I was stuck on this, Vitto helped me point out that I needed scrutor for DI.
I am not sure a map called "DomainEventHandlers" is the best fit is for DomainEventHandlers. 
I like Ardalis view to try to use namespaces to logically order files. As in `ExampleApp.Domain.GuestMessage` contains the `GuestMessageEvents, -EventHandlers, -Specifications.`
When using a Microsoft Library you'll see that the services, methods and extensions are more logically grouped in a Namespace that explain the domain.
e.g. `using Microsoft.Extensions.DependencyInjection;` to use the interface `IServiceCollection`.
It's not `using Microsoft.Extensions.Interfaces`.

That being said, it's not uncommon to group classes by type of the class, as now is implemented for domaineventhandler, I'm just not sure if it's a good practice.

**edit** when only using the interface of a microsoft library, it's common to use the e.g. using `Microsoft.Extensions.DependencyInjection.Abstractions;` package, but it's also grouped by logical domain.

**Comment by vittorelli** [comment]
> I modified it since I expect domainevents to be handled in the application layer in general. But I see now why you wrote it this way. A clever way to provide DI for methods on domain entity objects. I am not against it. I believe DomainEvents are maybe not supposed to be exclusively handled in the application layer.
> 
> @vittorelli What do you think?

Say you have more than one application using the same domain/infrastructure.
Would you expect the domain event handler to be executed in all applications, or only for a particular subset of applications?

Domain Event Handlers could be handled anywhere: in the domain itself, the application layer or infrastructure. I guess it depends on the nature of your event handler. The `GuestMessageCreatedEventHandler` seems to contain pure business logic, so I would expect it to be part of the domain.

A typical usage for my domain events are to keep aggregates independent of one another. When I run into trouble due to forbidden dependencies (i.e: the domain cannot depend on the application or infrastructure), I usually make an abstraction in the domain that is implemented in the infrastructure or application layer. For example, say you need to know the current user executing the core. In an ASP.NET application, this information would probably be made available through some authentication mechanism. An `ICurrentUserProvider` interface in the domain could then be implemented in the application layer and injected in domain event handlers.

Must Read: [Domain events: Design and implementation](https://learn.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/domain-events-design-implementation#domain-events-as-a-preferred-way-to-trigger-side-effects-across-multiple-aggregates-within-the-same-domain)

File: sample/Vivarni.Example.Domain/Entities/BaseEntity.cs
Comment (vittorelli) [inline sample/Vivarni.Example.Domain/Entities/BaseEntity.cs:1]: The data annotations seems more more like an infrastructure thing. Does the real world require the names of people to be maximum 50 characters? We can change this at a later point in time :-)

**Review by vittorelli — APPROVED** [review]

**Comment by vittorelli** [comment]
>I like Ardalis view to try to use namespaces to logically order files. As in ExampleApp.Domain.GuestMessage contains the GuestMessageEvents, -EventHandlers, -Specifications.

Totally agree. For the example however, we could make the argument that it is easier to understand what is going on as we only have a single aggregate at the moment. We could group everything in a namespace that reflects that aggregate (eg: GuestMessages), but that wouldn't be very helpfull for people navigating the current sample code. We might need to revisit if the sample is extended :-)

**Pull request merged**